File: 1term/10.28/1/1/list.cpp
#include "list.h"
#include <iostream>

using namespace std;

bool equals(ElementType value1, ElementType value2)
{
	return (value1 == value2);
}

namespace list
{
	struct ListElement
	{
		ElementType value;
		ListElement *next;
	};

	struct List
	{
		ListElement *head;
	};

	List *create()
	{
		List *result = new List;
		result->head = nullptr;
		return result;
	}

	Position createListElement(List *list, ElementType value)
	{
		ListElement *temp = new ListElement;
		temp->value = value;
		temp->next = nullptr;
		return temp;
	}

	void addElement(List *list, ElementType value, Position position)
	{
		ListElement *temp = new ListElement;
		temp->value = value;
		temp->next = position->next;
		position->next = temp;
	}

	void addElementToHead(List *list, ElementType value)
	{
		ListElement *temp = new ListElement;
		temp->value = value;
		temp->next = list->head;
		list->head = temp;
	}

	void deleteElement(List *list, Position position)
	{
		if (position == nullptr) 
			return;
		Position temp = position;
		if (position == list->head)
			list->head = list->head->next;
		else
		{
			position = list->head;
			while (position->next != temp)
				position = position->next;
			position->next = temp->next;
		}
		delete temp;
	}

	void deleteList(List *list)
	{
		if (list == nullptr)
			return;
		Position i = list->head;
		while (i != nullptr)
		{
			Position temp = i;
			i = i->next;
			delete temp;
		}
		delete list;
	}

	Position positionOfElement(List *list, ElementType value)
	{
		Position position = list->head;
		while (position != nullptr)
		{
			if (equals(position->value, value))
				return position;
			else
				position = position->next;
		}
		return nullptr;
	}

	ListElement *headListElement(List *list)
	{
		return list->head;
	}

	ListElement *nextListElement(ListElement *element)
	{
		return element->next;
	}

	ElementType valueListElement(List *list, ListElement *element)
	{
		return element->value;
	}

	void nextElementDoElement(List *list, Position temp, Position next)
	{
		if (next != nullptr)
			next->next = temp->next;
		temp->next = next;
	}

	void print(List *list)
	{
		if (list == nullptr || list->head == nullptr)
		{
			cout << "empty list" << endl;
			return;
		}
		Position temp = list->head;
		while (temp != nullptr)
		{
			cout << temp->value << " " ;
			temp = temp->next;
		}
		cout << endl;
	}	

	Position endOfList(List *list)
	{
		return nullptr;
	}

	int lengthList(List *list)
	{
		int i = 0;
		Position temp = headListElement(list);
		while (temp != NULL)
		{
			i++;
			temp = nextListElement(temp);
		}
		return i;
	}
}

File: 2term/02.19/3hash/3hash/List.cs
using System;
using ElementType = System.Int32;

namespace NList
{
    class List
    {
        public List()
        {
            head = null;
        }

        public void AddElement(ElementType value)
        {
            if (head == null)
            {
                head = new ListElement(value, null);
                return;
            }
            ListElement temp = head;
            for (temp = head; temp.Next() != null; temp = temp.Next()) ;
            temp.AddListElement(value);
        }

        public void AddElementToHead(ElementType value)
        {
            head = new ListElement(value, head);
        }

        public void DeleteElement(ElementType deleteNum)
        {
            if (head == null)
            {
                Console.WriteLine("List is empty");
                return;
            }
            ListElement temp = head;
            if (head.Value() == deleteNum)
                head = head.Next();
            else
            {
                while (temp.Next() != null && temp.Next().Value() != deleteNum)
                    temp = temp.Next();
                if (temp.Next() != null)
                    temp.DeleteListElement();
            }
        }

        public bool ExistList(ElementType value)
        {
            if (head == null)
                return false;
            ListElement temp = head;
            for (temp = head;  temp.Next() != null && temp.Value() != value; temp = temp.Next());
            return (temp.Value() == value);
        }

        public void Print()
        {
            Console.WriteLine();
            if (head == null)
            {
                Console.WriteLine("List is Empty");
                return;
            }
            ListElement temp = head;
            for (temp = head; temp.Next() != null; temp = temp.Next())
                Console.Write("{0}, ", temp.Value());
            Console.WriteLine("{0}", temp.Value());
        }

        private ListElement head;

        private class ListElement
        {
            public ListElement()
            { }

            public ListElement(ElementType value, ListElement next)
            {
                this.value = value;
                this.next = next;
            }

            public void AddListElement(ElementType value)
            {
                ListElement temp = new ListElement(value, this.next);
                this.next = temp;
            }

            // delete after this.
            public void DeleteListElement()
            {
                this.next = this.next.next;
            }

            public ElementType Value()
            {
                return this.value;
            }

            public ListElement Next()
            {
                return this.next;
            }

            private ElementType value;
            private ListElement next;
        }

    }
}
